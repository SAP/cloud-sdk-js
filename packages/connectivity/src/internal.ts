export * from './http-agent';
export * from './scp-cf';
