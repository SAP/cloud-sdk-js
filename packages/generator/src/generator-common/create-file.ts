import { join } from 'path';
import { promises } from 'fs';
import { codeBlock, ErrorWithCause, unixEOL } from '@sap-cloud-sdk/util';

const { writeFile } = promises;
/* eslint-disable valid-jsdoc */
/**
 * @internal
 */
export async function createFile(
  directoryPath: string,
  fileName: string,
  content: string,
  overwrite: boolean,
  withCopyright = true
): Promise<void> {
  try {
    return await writeFile(
      join(directoryPath, fileName),
      withCopyright ? wrapContent(content) : content,
      {
        encoding: 'utf8',
        flag: overwrite ? 'w' : 'wx'
      }
    );
  } catch (err) {
    const recommendation =
      err.code === 'EEXIST' && !overwrite
        ? ' File already exists. If you want to allow overwriting files, enable the `overwrite` flag.'
        : '';
    throw new ErrorWithCause(
      `Could not write file "${fileName}".${recommendation}`,
      err
    );
  }
}

function wrapContent(content: string): string {
  return (
    codeBlock`
${getCopyrightHeader()}
${content}
` + unixEOL
  );
}

function getCopyrightHeader(): string {
  return codeBlock`
/*
 * Copyright (c) ${new Date().getFullYear()} SAP SE or an SAP affiliate company. All rights reserved.
 *
 * This is a generated file powered by the SAP Cloud SDK for JavaScript.
 */
 `;
}
