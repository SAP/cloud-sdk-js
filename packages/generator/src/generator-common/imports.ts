import { codeBlock } from '@sap-cloud-sdk/util';
/**
 * @internal
 */
export interface Import {
  /**
   * Names of the exports that are to be imported.
   */
  names: string[];
  /**
   * The module to import from.
   */
  moduleIdentifier: string;
  /**
   * Indicates whether this is a type-only import.
   */
  typeOnly?: boolean;
}

/**
 * @internal
 */
export function serializeImports(imports: Import[]): string {
  const relevantImports = imports.filter(({ names }) => names.length);
  return relevantImports
    .map(
      ({ names, moduleIdentifier, typeOnly }) =>
        codeBlock`import ${typeOnly ? 'type ' : ''}{ ${names.join(
          ', '
        )} } from '${moduleIdentifier}';`
    )
    .join('\n');
}
