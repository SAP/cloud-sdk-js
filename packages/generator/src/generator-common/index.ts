export * from './create-file';
export * from './imports';
