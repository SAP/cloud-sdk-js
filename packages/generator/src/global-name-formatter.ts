import { propertyExists, UniqueNameGenerator } from '@sap-cloud-sdk/util';
import { ServiceMapping, VdmMapping } from './service-mapping';
/**
 * @internal
 */
export class GlobalNameFormatter {
  private directoryNameGenerator = new UniqueNameGenerator('-');
  private npmPackageNameGenerator = new UniqueNameGenerator('-');
  private vdmMapping: VdmMapping;

  constructor(vdmMapping: VdmMapping | undefined) {
    this.vdmMapping = vdmMapping || {};
    this.directoryNameGenerator.addToUsedNames(
      ...Object.entries(this.vdmMapping).map(([, v]) => v.directoryName)
    );
    this.npmPackageNameGenerator.addToUsedNames(
      ...Object.entries(this.vdmMapping).map(([, v]) => v.npmPackageName)
    );
  }

  public uniqueDirectoryName(
    directoryName: string,
    originalFileName: string
  ): string {
    return (
      this.directoryNameFromMapping(originalFileName) ||
      this.transformAndCacheDirectoryName(directoryName)
    );
  }

  public uniqueNpmPackageName(
    npmPackageName: string,
    originalFileName: string
  ): string {
    return (
      this.npmPackageNameFromMapping(originalFileName) ||
      this.transformAndCacheNpmPackageName(npmPackageName)
    );
  }

  private transformAndCacheDirectoryName(directoryName: string): string {
    return this.directoryNameGenerator.generateAndSaveUniqueName(directoryName);
  }

  private transformAndCacheNpmPackageName(npmPackageName: string): string {
    return this.npmPackageNameGenerator.generateAndSaveUniqueName(
      npmPackageName
    );
  }

  private directoryNameFromMapping(
    originalFileName: string
  ): string | undefined {
    return this.fromMapping(originalFileName, 'directoryName');
  }

  private npmPackageNameFromMapping(
    originalFileName: string
  ): string | undefined {
    return this.fromMapping(originalFileName, 'npmPackageName');
  }

  private fromMapping(
    originalFileName: string,
    key: keyof ServiceMapping
  ): string | undefined {
    return propertyExists(this.vdmMapping, originalFileName, key)
      ? this.vdmMapping[originalFileName][key]
      : undefined;
  }
}
