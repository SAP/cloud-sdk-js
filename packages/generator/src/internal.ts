export * from './operations';
export * from './aggregator-package';
export * from './batch';
export * from './complex-type';
export * from './edmx-parser';
export * from './edmx-to-vdm';
export * from './entity';
export * from './enum-type';
export * from './request-builder';
export * from './sdk-metadata';
export * from './service';
export * from './swagger-parser';
export * from './file-generator';
export * from './generator-options';
export * from './generator-utils';
export * from './generator';
export * from './global-name-formatter';
export * from './imports';
export * from './input-path-provider';
export * from './internal-prefix';
export * from './name-formatting-strategies';
export * from './options-parser';
export * from './reserved-words';
export * from './service-generator';
export * from './service-mapping';
export * from './service-name-formatter';
export * from './typedoc';
export * from './vdm-types';

// Had to be added for public API check
export * from './generator-without-ts-morph';
