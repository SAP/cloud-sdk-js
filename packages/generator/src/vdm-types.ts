import { PathLike } from 'fs';
import { ODataVersion } from '@sap-cloud-sdk/util';
/**
 * @internal
 */
export interface VdmServicePackageMetaData {
  /**
   * @internal
   */
  oDataVersion: ODataVersion;
  /**
   * @internal
   */
  namespaces: string[];
  /**
   * @internal
   */
  originalFileName: string;
  /**
   * @internal
   */
  servicePath: string;
  /**
   * @internal
   */
  directoryName: string;
  /**
   * @internal
   */
  npmPackageName: string;
  /**
   * @internal
   */
  speakingModuleName: string;
  /**
   * @internal
   */
  apiBusinessHubMetadata?: ApiBusinessHubMetadata | undefined;
  /**
   * @internal
   */
  className: string;
  /**
   * @internal
   */
  edmxPath: PathLike;
}
/**
 * @internal
 */
export interface VdmServiceEntities {
  /**
   * @internal
   */
  entities: VdmEntity[];
  /**
   * @internal
   */
  complexTypes: VdmComplexType[];
  /**
   * @internal
   */
  enumTypes: VdmEnumType[];
  /**
   * @internal
   */
  functionImports: VdmOperation[];
  /**
   * @internal
   */
  actionImports?: VdmOperation[];
}
/**
 * @internal
 */
export type VdmServiceMetadata = VdmServicePackageMetaData & VdmServiceEntities;

/**
 * Entity
 * @internal
 */
export interface VdmEntity {
  // todo uniqueness check like request builder
  // apiName: string;
  /**
   * @internal
   */
  entitySetName: string;
  /**
   * @internal
   */
  entityTypeName: string;
  /**
   * @internal
   */
  className: string;
  /**
   * @internal
   */
  creatable: boolean;
  /**
   * @internal
   */
  updatable: boolean;
  /**
   * @internal
   */
  deletable: boolean;
  /**
   * @internal
   */
  keys: VdmProperty[];
  /**
   * @internal
   */
  properties: VdmProperty[];
  /**
   * @internal
   */
  navigationProperties: VdmNavigationProperty[];
  /**
   * @internal
   */
  description: string;
  /**
   * @internal
   */
  entityTypeNamespace: string;
  /**
   * @internal
   */
  functions: VdmOperation[];
  /**
   * @internal
   */
  actions: VdmOperation[];
}

/**
 * Subset of Entity required for generating bound actions and functions.
 * This is required because at the time of generating bound actions and functions the full list of entities is not yet available.
 * But the properties included here are needed for generating the functions and actions.
 * @internal
 */
export type VdmPartialEntity = Pick<
  VdmEntity,
  'entityTypeName' | 'entityTypeNamespace' | 'className'
>;

/**
 * Properties
 * @internal
 */
export interface VdmPropertyDescriptor {
  /**
   * @internal
   */
  originalName: string;
  /**
   * @internal
   */
  instancePropertyName: string;
  /**
   * @internal
   */
  propertyNameAsParam: string;
  /**
   * @internal
   */
  staticPropertyName: string;
  /**
   * @internal
   */
  isComplex?: boolean;
  /**
   * @internal
   */
  isEnum?: boolean;
  /**
   * @internal
   */
  isCollection: boolean;
}
/**
 * @internal
 */
export interface VdmPropertyValueConstraints {
  /**
   * @internal
   */
  maxLength?: string;
  /**
   * @internal
   */
  nullable: boolean;
}
/**
 * @internal
 */
export interface VdmProperty
  extends VdmPropertyDescriptor,
    VdmPropertyValueConstraints,
    VdmMappedEdmType {
  /**
   * @internal
   */
  description: string;
}

/**
 * Navigation Properties
 * @internal
 */
export interface VdmNavigationProperty extends VdmPropertyDescriptor {
  /**
   * @internal
   */
  from: string;
  /**
   * @internal
   */
  to: string;
  /**
   * @internal
   */
  toEntityClassName: string;
}
/**
 * @internal
 */
export interface VdmEnumMemberType {
  /**
   * @internal
   */
  name: string;
  /**
   * @internal
   */
  originalValue: string;
}
/**
 * @internal
 */
export interface VdmEnumType {
  /**
   * @internal
   */
  originalName: string;
  /**
   * @internal
   */
  typeName: string;
  /**
   * @internal
   */
  underlyingType: string;
  /**
   * @internal
   */
  members: VdmEnumMemberType[];
}

/**
 * Complex types and function imports
 * @internal
 */
export interface VdmComplexType {
  /**
   * @internal
   */
  originalName: string;
  /**
   * @internal
   */
  properties: VdmProperty[];
  /**
   * @internal
   */
  typeName: string;
  /**
   * @internal
   */
  fieldType: string;
  /**
   * @internal
   */
  namespace: string;
}
/**
 * @internal
 * Represents a partial operation (action or function).
 */
export interface VdmOperationBase {
  /**
   * @internal
   */
  originalName: string;
  /**
   * @internal
   */
  parameters: VdmParameter[];
  /**
   * @internal
   */
  parametersTypeName: string;
  /**
   * @internal
   */
  name: string;
  /**
   * @internal
   */
  description: string;
  /**
   * @internal
   */
  type: 'function' | 'action';
  /**
   * @internal
   */
  isBound: boolean;
  /**
   * @internal
   * only set for bound operations
   */
  entityClassName?: string;
}

/**
 * @internal
 * Represents either a function or an action.
 */
export interface VdmOperation extends VdmOperationBase {
  /**
   * @internal
   */
  httpMethod: string;
  /**
   * @internal
   */
  returnType: VdmOperationReturnType;
}

/**
 * @internal
 */
export interface VdmOperationReturnType {
  /**
   * @internal
   */
  builderFunction?: string;
  /**
   * @internal
   */
  returnType: string;
  /**
   * @internal
   */
  isCollection: boolean;
  /**
   * @internal
   */
  isNullable: boolean;
  /**
   * @internal
   */
  returnTypeCategory: VdmReturnTypeCategory;
  /**
   * @internal
   */
  unsupportedReason?: VdmUnsupportedReason;
}
/**
 * @internal
 */
export enum VdmReturnTypeCategory {
  ENTITY,
  COMPLEX_TYPE,
  EDM_TYPE,
  VOID,
  NEVER
}
/**
 * @internal
 */
export enum VdmUnsupportedReason {
  ENTITY_NOT_DESERIALIZABLE
}
/**
 * @internal
 */
export interface VdmParameter extends VdmMappedEdmType {
  /**
   * @internal
   */
  originalName: string;
  /**
   * @internal
   */
  parameterName: string;
  /**
   * @internal
   */
  nullable: boolean;
  /**
   * @internal
   */
  description: string;
}
/**
 * @internal
 */
export interface VdmMappedEdmType {
  /**
   * @internal
   */
  edmType: string;
  /**
   * @internal
   */
  jsType: string;
  /**
   * @internal
   */
  fieldType: string;
}
/**
 * @internal
 */
export interface ApiBusinessHubMetadata {
  /**
   * @internal
   */
  url: string;
  /**
   * @internal
   */
  communicationScenario: string | null;
  /**
   * @internal
   */
  businessDocumentationUrl?: string;
}
