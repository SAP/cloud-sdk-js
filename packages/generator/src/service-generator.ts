import { ParsedGeneratorOptions } from './generator-options';
import { npmCompliantName } from './generator-utils';
import { GlobalNameFormatter } from './global-name-formatter';
import { inputPaths, ServiceDefinitionPaths } from './input-path-provider';
import {
  getServicePath,
  readServiceMapping,
  VdmMapping
} from './service-mapping';
import { ServiceNameFormatter } from './service-name-formatter';
import { readEdmxFile, ServiceMetadata } from './edmx-parser/edmx-file-reader';
import { readSwaggerFile } from './swagger-parser/swagger-parser';
import { apiBusinessHubMetadata } from './swagger-parser/swagger-util';
import { VdmServiceMetadata, VdmServicePackageMetaData } from './vdm-types';
import { isV2Metadata } from './edmx-to-vdm/edmx-to-vdm-util';
import { getServiceEntitiesV2 } from './edmx-to-vdm/v2';
import { getServiceEntitiesV4 } from './edmx-to-vdm/v4';

class ServiceGenerator {
  private globalNameFormatter: GlobalNameFormatter;
  private serviceMapping: VdmMapping;

  constructor(readonly options: ParsedGeneratorOptions) {
    this.serviceMapping = readServiceMapping(options);
    this.globalNameFormatter = new GlobalNameFormatter(this.serviceMapping);
  }

  public withServiceMapping(serviceMapping: VdmMapping) {
    this.serviceMapping = serviceMapping;
    return this;
  }

  public withGlobalNameFormatter(globalNameFormatter: GlobalNameFormatter) {
    this.globalNameFormatter = globalNameFormatter;
    return this;
  }

  public generateAllServices(): VdmServiceMetadata[] {
    return inputPaths(this.options.inputDir, this.options.useSwagger).map(p =>
      this.generateService(p)
    );
  }

  public generateService(
    serviceDefinitionPaths: ServiceDefinitionPaths
  ): VdmServiceMetadata {
    const serviceMetadata = this.readEdmxAndSwaggerFile(serviceDefinitionPaths);

    const vdmServicePackageMetaData = this.getServicePackageMetaData(
      serviceMetadata,
      serviceDefinitionPaths
    );

    const vdmServiceEntities = isV2Metadata(serviceMetadata.edmx)
      ? getServiceEntitiesV2(
          serviceMetadata,
          vdmServicePackageMetaData.className
        )
      : getServiceEntitiesV4(
          serviceMetadata,
          vdmServicePackageMetaData.className
        );

    return {
      ...vdmServicePackageMetaData,
      ...vdmServiceEntities
    };
  }

  private getServicePackageMetaData(
    serviceMetadata: ServiceMetadata,
    serviceDefinitionPaths: ServiceDefinitionPaths
  ): VdmServicePackageMetaData {
    const directoryName = this.globalNameFormatter.uniqueDirectoryName(
      ServiceNameFormatter.originalToServiceName(
        serviceMetadata.edmx.namespaces.length === 1
          ? serviceMetadata.edmx.namespaces[0]
          : serviceMetadata.edmx.fileName
      ),
      serviceMetadata.edmx.fileName
    );
    const npmPackageName = this.globalNameFormatter.uniqueNpmPackageName(
      npmCompliantName(directoryName),
      serviceMetadata.edmx.fileName
    );
    const speakingModuleName =
      ServiceNameFormatter.directoryToSpeakingModuleName(directoryName);
    const className = `${speakingModuleName.replace(/ /g, '')}`;

    return {
      oDataVersion: serviceMetadata.edmx.oDataVersion,
      namespaces: serviceMetadata.edmx.namespaces,
      originalFileName: serviceMetadata.edmx.fileName,
      directoryName,
      npmPackageName,
      speakingModuleName,
      servicePath: getServicePath(
        serviceMetadata,
        this.serviceMapping[serviceMetadata.edmx.fileName]
      ),
      edmxPath: serviceDefinitionPaths.edmxPath,
      apiBusinessHubMetadata: apiBusinessHubMetadata(serviceMetadata.swagger),
      className
    };
  }

  private readEdmxAndSwaggerFile(
    serviceDefinitionPaths: ServiceDefinitionPaths
  ): ServiceMetadata {
    const serviceMetadata: ServiceMetadata = {
      edmx: readEdmxFile(serviceDefinitionPaths.edmxPath)
    };
    if (serviceDefinitionPaths.swaggerPath) {
      serviceMetadata.swagger = readSwaggerFile(
        serviceDefinitionPaths.swaggerPath
      );
    }
    return serviceMetadata;
  }
}

/**
 * @param options - Generator options
 * @returns the parsed services
 * @internal
 */
export function parseAllServices(
  options: ParsedGeneratorOptions
): VdmServiceMetadata[] {
  return new ServiceGenerator(options).generateAllServices();
}

/**
 * @param serviceDefinitionPaths - Path to the service definition
 * @param options - Generator options
 * @param mappings - mappings for VDM service names to desired name
 * @param globalNameFormatter - Instance of global name formatter to be used for the parsing process
 * @returns the parsed service
 * @internal
 */
export function parseService(
  serviceDefinitionPaths: ServiceDefinitionPaths,
  options: ParsedGeneratorOptions,
  mappings: VdmMapping,
  globalNameFormatter: GlobalNameFormatter
): VdmServiceMetadata {
  return new ServiceGenerator(options)
    .withServiceMapping(mappings)
    .withGlobalNameFormatter(globalNameFormatter)
    .generateService(serviceDefinitionPaths);
}

/**
 * @internal
 */
export function getServiceName(service: VdmServiceMetadata): string {
  return service.namespaces.length === 1
    ? service.namespaces[0]
    : service.originalFileName;
}
