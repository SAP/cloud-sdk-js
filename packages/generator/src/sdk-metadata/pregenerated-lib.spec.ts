import { resolve } from 'path';
import { packageDescription } from '@sap-cloud-sdk/generator-common/internal';
import { VdmServiceMetadata } from '../vdm-types';
import { parseService } from '../service-generator';
import { createOptions } from '../../test/test-util/create-generator-options';
import { GlobalNameFormatter } from '../global-name-formatter';
import { oDataServiceSpecs } from '../../../../test-resources/odata-service-specs';
describe('pregenerated-lib', () => {
  const service: VdmServiceMetadata = getTestService();

  it('returns description of the service', () => {
    expect(packageDescription(service.speakingModuleName)).toMatchSnapshot();
  });
});

export function getTestService(npmPackageName?: string): VdmServiceMetadata {
  return parseService(
    {
      edmxPath: resolve(
        oDataServiceSpecs,
        'v2',
        'API_TEST_SRV/API_TEST_SRV.edmx'
      )
    },
    createOptions(),
    {},
    new GlobalNameFormatter({
      API_TEST_SRV: {
        directoryName: 'test-service',
        servicePath: '/sap/opu/odata/sap/API_TEST_SERVICE_SRV;v=0002',
        npmPackageName: npmPackageName || '@sap/cloud-sdk-vdm-test-service'
      }
    })
  );
}
