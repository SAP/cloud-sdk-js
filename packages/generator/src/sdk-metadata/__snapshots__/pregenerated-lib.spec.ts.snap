// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pregenerated-lib returns description of the service 1`] = `"SAP Cloud SDK for JavaScript: Generated client for service Test Service"`;
